File: src/game/PlayerBullet.cpp
/*
 *  This file is part of Last Escape.
 *
 *  Last Escape is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  Last Escape is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Last Escape.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "PlayerBullet.h"
#include "Player.h"
#include "Enemy.h"
#include "globals.h"

PlayerBullet::PlayerBullet(double x, double y, double w, double h, int facing, double lifetime):
AnimatedActor(x, y, w, h)
{
	this->lifetime = lifetime;
	bulletTime = 0;

	facing_direction = facing;

	// Put it in the PLayerBullets group so it doesn't collide with other PlayerBullets
	shape->group = PhysicsGroup::PlayerBullets;
	shape->layers = PhysicsLayer::Map|PhysicsLayer::PlayerBullet;
	shape->collision_type = PhysicsType::PlayerBullet;

	canSleep = false;
}

void PlayerBullet::collide(Actor& otherActor) {
	if (otherActor.isEnemy())
	{
		dynamic_cast<Enemy * >(&otherActor)->doDamage(damage);
		//if(dynamic_cast<Enemy * >(&otherActor)->doDamage(damage)) {
			//g_player->addExperience(otherActor.getExperienceValue());
		//}
		die();
	}
	else if ( dynamic_cast<PhysicsCrate*>(&otherActor) != NULL )
	{
		((PhysicsCrate&)otherActor).doDamage(damage);
		die();
	}
}

void PlayerBullet::update(sf::Uint32 dt) {
	//cout << "bullet " << body->p.x << " " << body->p.y << "\n";
	bulletTime += dt;
	//cout << "angle " << rad2deg(body->a) << "\n";

	updateSpriteFacing();
	//checkcollisions();
	if(bulletTime > lifetime) {
		destroy();
	}
}

File: scripts/playerBullet.js
function playerBulletUpdate(dt) {
	this.bulletTime += dt;
	//cout << "angle " << rad2deg(body->a) << "\n";

	this.updateSpriteFacing();
	//checkcollisions();
	if(this.bulletTime > this.lifetime) {
		this.destroy();
	}
}


function playerBulletCollide(otherActor)
{
	if (otherActor.shape.collision_type == PhysicsType.enemy) {
		otherActor.doDamage(damage);
		this.destroy();
	}
}

function playerBulletDestroy()
{
	this.destroy();
}

function playerBullet(x, y) {
	var a = new Actor(x, y, "playerBullet.xml");
	a.actorName = "playerBullet";
	a.speed = 500.0;
	a.angle = ((a.facing == Facing.right) ? 90 : -90) + (frand(100, 200)) * angleVariation/100;
	a.speed_x = 0.5 + Math.sin(a.angle * 3.14159/180.0) * a.speed;
	a.speed_y = 0.5 + Math.cos(a.angle * 3.14159/180.0) * a.speed;

	// Put it in the PLayerBullets group so it doesn't collide with other PlayerBullets
	a.collisionGroup = PhysicsGroup.playerBullets;
	a.shapeLayers = PhysicsLayer.map|PhysicsLayer.playerBullet;
	a.collisionType = PhysicsType.playerBullet;

	a.update = playerBulletUpdate;
	a.onCollide = playerBulletCollude;
	a.onCollideGround = playerBulletDestroy;
	a.onCollideWall = playerBulletDestroy;

	a.body.velocity.x = a.speed_x;
	a.body.velocity.y = a.speed_y;

	if(a.facing == Facing.right)
		a.body.rotationDegrees = -a.angle - 270;
	else
		a.body.rotationDegrees = -a.angle - 90;

	//TODO
	//setVelocityFunc(no_gravity);

	a.damage = 1;
	a.canSleep = false;

	a.animation = "bullet";
}
